File: lib/core/services/chart_data_builder.dart
import '../models/chart_dataset.dart';
import '../models/grade_config.dart';
import '../models/processing_report.dart';

class ChartDataBuilder {
  const ChartDataBuilder();

  ChartDataset buildGradeDistribution(ProcessingReport report) {
    final points = report.summary.gradeCounts.entries
        .where((entry) => entry.value > 0)
        .map((entry) => ChartPoint(label: entry.key.label, count: entry.value))
        .toList();

    return ChartDataset(points: points);
  }
}

File: Assignment/Chapter 1/Student Grade Calculator/Kotlin Version/app/src/main/java/com/romualdsigning/studentgradecalc/domain/usecase/ChartDataBuilder.kt
package com.romualdsigning.studentgradecalc.domain.usecase

import com.romualdsigning.studentgradecalc.domain.model.ChartDataset
import com.romualdsigning.studentgradecalc.domain.model.ChartPoint
import com.romualdsigning.studentgradecalc.domain.model.ProcessingReport

class ChartDataBuilder {
    fun buildGradeDistribution(report: ProcessingReport): ChartDataset {
        // I convert the raw grade-count map once here so the UI only renders chart-ready points.
        val points = report.summary.gradeCounts
            .filterValues { it > 0 }
            .map { (grade, count) -> ChartPoint(label = grade.label, count = count) }

        return ChartDataset(points = points)
    }
}

